### DISCUSSION TITLE
How to return custom data/error on action

### DISCUSSION THREAD

**Discussion by ronkristoff** [Code-Forge-Net/remix-hook-form#7] [Q&A] [answered]
currently  i have this on my action function,

> export const action = async ({ request }: ActionArgs) => {
  const {
    errors,
    data,
    receivedValues: defaultValues,
  } = await getValidatedFormData<FormData>(request, resolver);
  if (errors) {
    return json({ errors, defaultValues });
  }
  //return some data that comes from custom validation
  return json({ customErrors: "test" });
  
  
im just trying to simulate a scenario where in i need to do a custom validation on the action function, but when i return a new property i get an error on useActionData.

> 'customErrors' is declared but its value is never read.ts(6133)
Property 'customErrors' does not exist on type 'SerializeObject<UndefinedToOptional<{ title: string; complete: boolean; }>> | SerializeObject<UndefinedToOptional<{ errors: FieldErrors<{ title: string; complete: boolean; }>; defaultValues: Record<...>; }>> | SerializeObject<...> | undefined'.ts(2339)

how do i resolve this? or what is the best way to deal with this scenario? Thanks.

**Comment by AlemTuzlak** [comment, marked as answer]
I think your main issue here @ronkristoff  is that your
```jsx
export const action = async ({ request }: ActionArgs) => {
  const {
    errors,
    data,
    receivedValues: defaultValues,
  } = await getValidatedFormData(request, resolver);
  if (errors) {
     return json({ errors, defaultValues });
  }
  //return some data that comes from custom validation
  return json({ customErrors: "test" });
```
is returning either errors and defaultValues or customErrors, if you try doing it like this:
```jsx
export const action = async ({ request }: ActionArgs) => {
  const {
    errors,
    data,
    receivedValues: defaultValues,
  } = await getValidatedFormData(request, resolver);
  if (errors) {
     return json({ errors, defaultValues, customErrors: "" });
  }
  //return some data that comes from custom validation
  return json({ customErrors: "test" });
```
it should make typescript happy because you do not return the customErrors on every request typescript can't guarantee that it will be there

**Comment by Centerworx** [comment]
On the topic of errors, should custom errors conform to the  { fieldName: { message: string, type?: string } } type to be added as errors, since this is the error pattern from react-hook-forms?
